txnsync/bitmask.go

```diff
@@ -21,29 +21,51 @@ import (
 	"errors"
 )
 
+var errIndexOutOfBounds = errors.New("invalid bitmask: index out of bounds")
+var errIndexNotFound = errors.New("invalid bitmask: index not found")
+var errInvalidBitmaskType = errors.New("invalid bitmask type")
+
 //msgp:allocbound bitmask maxBitmaskSize
 type bitmask []byte
 
 // assumed to be in mode 0, sets bit at index to 1
-func (b *bitmask) SetBit(index int) {
+func (b *bitmask) setBit(index int) error {
 	byteIndex := index/8 + 1
+	if index < 0 || byteIndex >= len(*b) {
+		return errIndexOutOfBounds
+	}
 	(*b)[byteIndex] |= 1 << (index % 8)
+	return nil
 }
 
-func (b *bitmask) EntryExists(index int, entries int) bool {
+// entryExists converts the bitmask to type 0 (if not already)
+func (b *bitmask) entryExists(index int, entries int) bool {
 	if len(*b) == 0 {
 		return false
 	}
 	if (*b)[0] != 0 {
-		b.expandBitmask(entries)
+		err := b.expandBitmask(entries)
+		if err != nil {
+			return false
+		}
 	}
 	byteIndex := index/8 + 1
 	return byteIndex < len(*b) && ((*b)[byteIndex]&(1<<(index%8)) != 0)
 }
 
-func (b *bitmask) trimBitmask(entries int) {
+// trimBitmask compresses the bitmask into one of the 4 types:
+// type 0: intput bitmask bit pos x b -> output bitmask bit pos x b
+// type 1: intput bitmask bit pos x b -> output bitmask bit pos x !b
+// type 2: stores the positions of bits where b = 1
+//         intput bitmask first b=1 pos A, second b=1 pos B, ...
+//         output bitmask byte 2,A/256,A%256,(B-A)/256,(B-A)%256,...
+// type 3: same as type 2, but stures the positons where b = 0
```

Comment:
typos: `same as type 2, but stures the positons where b = 0` -> `same as type 2, but stores the positions where b = 0`